File: GCons/B2.m
classdef B2
    %B2 constraint using r-p formulation. The B2 is an intermediate level
    %constraint which exists between a plane, formed by 2 vectors aiBar and
    %biBar on body i, and a vector dij from point Pi to point Qj
    %(note: see slide 24 from 9/26 for details and notation)
    %the constraint is made using 2 dp2 constraints, and removes 2 degrees
    %of freedom by enforcing 2 ACE's
  
    
    properties
        DOFremoved = 2;  %DOF removed
        bodyi; %instance of a body object, i 
        bodyj; %instance of a body object, j
        dp2_a;%instance of dp2 constraint between aiBar and dij
        dp2_b;%instance of dp2 constraint between biBar and dij
        
        %these point definitions are taken from S24 lecture 9.26
        aiHead; %index of a [3x1] marker contained within bodyi, representing head of aiBar
        aiTail; %index of a [3x1] marker contained within bodyj, representing tail of aiBar
        biHead; %index of a [3x1] marker contained within bodyi, representing head of aiBar
        biTail; %index of a [3x1] marker contained within bodyj, representing tail of aiBar 
        Pi;     %index of a [3x1] marker contained within bodyi, representing point Pi
        Qj;     %index of a [3x1] marker contained within bodyj, representing Qj
          
        
    end
    
    %calculated values 
    properties(Dependent)
        phi;   %constraint equation             [2x1] 
        nu;    %-dPhi/ dt                       [2x1]
        gamma; %RHS of acceleration equation    [2x1]
        
        phi_ri;%derivative of phi wrt ri        [2x3]  or [ ] if  i is ground
        phi_rj;%derivative of phi wrt rj        [2x3]  or [ ] if  j is ground
        phi_pi;%derivative of phi wrt ri        [2x4]  or [ ] if  i is ground
        phi_pj;%derivative of phi wrt ri        [2x4]  or [ ] if  j is ground

    end
    
    methods
        %constructor
        function b2 = B2(bodyi,bodyj,aiHead,aiTail,biHead,biTail,Pi,Qj)
            
            %bodies
            b2.bodyi = bodyi;
            b2.bodyj = bodyj;
            
            %vector head and tail indices
            b2.aiHead = aiHead;
            b2.aiTail = aiTail;
            b2.biHead = biHead;
            b2.biTail = biTail;
            b2.Pi  = Pi;
            b2.Qj = Qj;
           
            %form basic Gcons which make up B2 
            b2.dp2_a = DP2(b2.bodyi,b2.bodyj,b2.Pi,b2.Qj, b2.aiHead, b2.aiTail);
            b2.dp2_b = DP2(b2.bodyi,b2.bodyj,b2.Pi,b2.Qj, b2.biHead, b2.biTail);
            
            
        end
        
%--------------------getters for dependent properties-------------------------
        
        function phi = get.phi(b2)
            %phi [2x1] constraint equation
            phi = [b2.dp2_a.phi , b2.dp2_b.phi]';
        end
        
        function nu = get.nu(b2) 
           %nu [2x1] velocity equation [nu(1) , nu(2)]';
           nu = [b2.dp2_a.nu , b2.dp2_b.nu]';
        end
        
        function gamma = get.gamma(b2)
            %gamma [2x1] is the RHS of the acceleration equation
            gamma = [b2.dp2_a.gamma , b2.dp2_b.gamma]';
        end
       
      
        
        function phi_ri = get.phi_ri(b2)
            %phi_ri [2x3] is the derivative of both phi equations, with respect
            %to the positional coordinates of body i, if body i is ground,
            %phi_ri is []
            phi_ri = [b2.dp2_a.phi_ri, b2.dp2_b.phi_ri]';
          
        end
        
         function phi_rj = get.phi_rj(b2)
            %phi_rj [2x3] is the derivative of both phi equations, with respect
            %to the positional coordinates of body i, if body i is ground,
            %phi_rj is []
            phi_rj = [b2.dp2_a.phi_rj, b2.dp2_b.phi_rj]';
          
         end
        
         function phi_pi = get.phi_pi(b2)
            %phi_pi [2x3] is the derivative of both phi equations, with respect
            %to the positional coordinates of body i, if body i is ground,
            %phi_pi is []
            phi_pi = [b2.dp2_a.phi_pi, b2.dp2_b.phi_pi]';
          
         end
        
          function phi_pj = get.phi_pj(b2)
            %phi_pj [2x3] is the derivative of both phi equations, with respect
            %to the positional coordinates of body i, if body i is ground,
            %phi_pj is []
            phi_pj = [b2.dp2_a.phi_pj, b2.dp2_b.phi_pj]';
          
          end
        
       
    end
end
    

File: src/GCons/b2.jl
#GCons are Geometic Constraints
type b2
  """
  The b2 constraint specifies that a vector PiQj is perpendicular to a plane
  defined by two vectors on bodyi , ai and bi
  b1 is one of two intermediate constraints
  """
  sim::Sim      #reference to the simulation data structures
  bodyi::Body   #bodyi
  bodyj::Body   #bodyj
  Pi::Int       #index of head of vector Pi
  Qj::Int       #index of head of vector Qj
  ai_head::Int  #index of ai_head
  ai_tail::Int  #index of a1_tail
  bi_head::Int  #index of bi_head
  bi_tail::Int  #index of bi_tail
  rDOF::Int     #number of degrees of freedom removed by constraint
  subGCs::Array #consider using inhepitance here for speedup?


  #constructor function
  function b2(sim::Sim,bodyi::Body,bodyj::Body,Pi,Qj, ai_head, bi_head, ai_tail = 1, bi_tail = 1)
    rDOF = 2;

    subGCs = Array(Any,2)
    subGCs[1] = dp2(sim,bodyi,bodyj,Pi,Qj,ai_head,ai_tail)
    subGCs[2] = dp2(sim,bodyi,bodyj,Pi,Qj,bi_head,bi_tail)

    new(sim,bodyi,bodyj,Pi,Qj,ai_head,ai_tail,bi_head,bi_tail,rDOF,subGCs)
  end
end

#----------------begin functions associated with b2----------------------------

function ϕ(con::b2)   #9.26.2016 - slide 24
  """
  constraint equation ϕ
  output: [2 x 1] evaluation of constraint equation value
  """
  phi = [ ϕ(con.subGCs[1]) ; ϕ(con.subGCs[2])]
end

function ν(con::b2)
  """
  RHS of vel equation
  output: [2 x 1] evaluation ν
  """
  nu = [ ν(con.subGCs[1]) ; ν(con.subGCs[2])]
end

function 	𝛾(con::b2)  #10.7.2016 - slide 8
  """
  RHS of accel equation
  output: [2 x 1] evaluation ν
  """
  gamma = [ 𝛾(con.subGCs[1]) ; 𝛾(con.subGCs[2])]
end

function ϕ_r(con::b2)  #9.28.2016 slide 17
  """
  partial derivative of ϕ WRT position position GC's of both bodyi and bodyj
  output: ([2x3],[2x3])
  """
  phi_r = Array{Float64}(con.rDOF,6)
  phi_r[1:1,1:3], phi_r[1:1,4:6] = ϕ_r(con.subGCs[1])
  phi_r[2:2,1:3], phi_r[2:2,4:6] = ϕ_r(con.subGCs[2])
  return phi_r[:,1:3] , phi_r[:,4:6]
end

function ϕ_p(con::b2)  # #9.28.2016 slide 17
  """
  partial derivative of ϕ WRT position orientation GC's of both bodyi and bodyj
  output:([2x4],[2x4])
  """
  phi_p = Array{Float64}(con.rDOF, 8)
  phi_p[1:1,1:4], phi_p[1:1,5:8] = ϕ_p(con.subGCs[1])
  phi_p[2:2,1:4], phi_p[2:2,5:8] = ϕ_p(con.subGCs[2])
  return phi_p[:,1:4] , phi_p[:,5:8]
end
